ouroboros-consensus-test/bench-mempool/Bench/Consensus/Mempool/TestBlock.hs

```diff
@@ -87,29 +100,51 @@ sampleMempoolAndModelParams = MempoolAndModelParams {
   Payload semantics
 -------------------------------------------------------------------------------}
 
-data TestLedgerState = TestLedgerState {
-    availableTokens :: Set Token
-  }
-  deriving stock (Generic, Eq, Show)
-  deriving anyclass (NoThunks, ToExpr, Serialise)
-
-data TxApplicationError =
+newtype TxApplicationError =
     -- | The transaction could not be applied due to the given unavailable tokens.
     TxApplicationError { unavailable :: Set Token }
   deriving stock (Generic, Eq, Show)
   deriving anyclass (NoThunks, ToExpr, Serialise)
 
 instance PayloadSemantics Tx where
-  type PayloadDependentState Tx = TestLedgerState
+  newtype instance PayloadDependentState Tx mk = TestPLDS {
+      getTestPLDS :: mk Token ()
+    }
+    deriving stock Generic
 
   type PayloadDependentError Tx = TxApplicationError
 
-  applyPayload st@TestLedgerState { availableTokens } Tx { consumed, produced } =
-    let
-      notFound = Set.filter (not . (`Set.member` availableTokens)) consumed
-    in if Set.null notFound
-       then Right $ st{ availableTokens = availableTokens \\ consumed <> produced }
-       else Left  $ TxApplicationError notFound
+  applyPayload plds tx =
+      let
+        notFound = Set.filter (not . (`Map.member` tokMap)) consumed
+      in if Set.null notFound
+        then Right $ TestPLDS (Ledger.rawAttachAndApplyDiffs fullDiff toks)
+        else Left  $ TxApplicationError notFound
+    where
+      TestPLDS toks@(ValuesMK tokMap) = plds
+      Tx {consumed, produced}                = tx
+
+      consumedDiff, producedDiff :: Diff Token ()
+      consumedDiff = Diff.fromListDeletes [(t, ()) | t <- Set.toList consumed]
+      producedDiff = Diff.fromListInserts [(t, ()) | t <- Set.toList produced]
+
+      fullDiff :: DiffMK Token ()
+      fullDiff = DiffMK $ consumedDiff <> producedDiff
+
+  getPayloadKeySets tx = TestLedgerTables $ KeysMK $ consumed <> produced
+    where
+      Tx {consumed, produced} = tx
+
+deriving stock instance (IsMapKind mk)
+                        => Eq (PayloadDependentState Tx mk)
+deriving stock instance (IsMapKind mk)
+                        => Show (PayloadDependentState Tx mk)
+deriving anyclass instance (IsMapKind mk)
+                        => NoThunks (PayloadDependentState Tx mk)
+
+instance Serialise (PayloadDependentState Tx EmptyMK) where
+  encode = error "unused: encode"
+  decode = error "unused: decode"
```

Comment:
Why do we need this instance?

Reply:
`PayloadSemantics` requires it as a constraint

Reply:
We could try to remove it if it's not much work. Sorry if I introduced it necessarily :(

Reply:
Maybe I don't recall, hence the *if* :)